js/bitflyer.js

```diff
@@ -101,58 +101,133 @@ module.exports = class bitflyer extends Exchange {
+        //        "market_type": "Futures"
+        //    },
+        //
         const us_markets = await this.publicGetGetmarketsUsa (params);
+        //
+        //    {"product_code": "BTC_USD", "market_type": "Spot"},
+        //    {"product_code": "BTC_JPY", "market_type": "Spot"}
+        //
         const eu_markets = await this.publicGetGetmarketsEu (params);
+        //
+        //    {"product_code": "BTC_EUR", "market_type": "Spot"},
+        //    {"product_code": "BTC_JPY", "market_type": "Spot"}
+        //
         let markets = this.arrayConcat (jp_markets, us_markets);
         markets = this.arrayConcat (markets, eu_markets);
         const result = [];
         for (let i = 0; i < markets.length; i++) {
             const market = markets[i];
             const id = this.safeString (market, 'product_code');
             const currencies = id.split ('_');
-            let baseId = undefined;
-            let quoteId = undefined;
-            let base = undefined;
-            let quote = undefined;
-            const numCurrencies = currencies.length;
-            if (numCurrencies === 1) {
-                baseId = id.slice (0, 3);
-                quoteId = id.slice (3, 6);
-            } else if (numCurrencies === 2) {
-                baseId = currencies[0];
-                quoteId = currencies[1];
-            } else {
-                baseId = currencies[1];
-                quoteId = currencies[2];
-            }
-            base = this.safeCurrencyCode (baseId);
-            quote = this.safeCurrencyCode (quoteId);
-            const symbol = (numCurrencies === 2) ? (base + '/' + quote) : id;
-            const fees = this.safeValue (this.fees, symbol, this.fees['trading']);
-            let maker = this.safeValue (fees, 'maker', this.fees['trading']['maker']);
-            let taker = this.safeValue (fees, 'taker', this.fees['trading']['taker']);
-            let spot = true;
-            let future = false;
+            const marketType = this.safeString (market, 'market_type');
+            let swap = (marketType === 'FX');
+            let future = (marketType === 'Futures');
+            let spot = !swap && !future;
             let type = 'spot';
-            if (('alias' in market) || (currencies[0] === 'FX')) {
+            let settle = undefined;
+            let baseId = currencies[0];
+            let quoteId = currencies[1];
+            let expiry = undefined;
+            if (swap) {
+                type = 'swap';
+                baseId = currencies[1]
+                quoteId = currencies[2]
+            } else if (future) {
+                const alias = this.safeString(market, 'alias');
+                const currencyId = this.safeString (alias.split('_'), 0);
+                baseId = currencyId.slice (0,-3);
+                quoteId = currencyId.slice (-3);
+                const date = this.safeString (id.split(currencyId), 1);
+                const rfc2616 = this.rfc2616(date);
```

Comment:
We don't have an RFC 2616 parser yet. It also has a few linting issues (missing spaces mostly).

Reply:
Is there a method that will work that isn't this?

Reply:
The base `rfc2616` does not really parse anything, it takes a timestamp in milliseconds and converts it to a rfc2616 string. We can probably use parseDate on it, or we could break down the date in parts.

Reply:
`parseDate` didn't work, I looked through the time functions and this was the only one that worked

Reply:
The date variable has the format `11FEB2022`

Reply:
I'll try to resolve that.